### PR TITLE
#6 - CRUD for providers

### LINKED ISSUE #6: [Administrator panel] CRUD for providers
Administrator should be able to manage providers. Each providers should have data like name, logo, url to website. He should see information in which cities the provider is available and also has possibility to change this.

### LINKED ISSUE #96: Providers logos should have same sizes and proportions
Right now we have a huge mess with images of providers:
* Bird is 150x100px png image
* Helbiz is 400x61px png
* Zwings is 1136 x150px (!)

I think you should discuss if they should have transparent backgroud.

AC:
* all images have the same sizes, proportions and formats
* there is a test case to check if criterion above is green

### PR DESCRIPTION
**Pull request by AleksandraKozubal**
Admin can see all the providers with its' logo, color and url. Delete is also possible (deleted provider is gone until next seed).
image

Create and edit buttons are done, so I left it as 'disabled' option, for later use.

Creating and editing provider is not so easy to do, coz all the info are seeded - we should rethink and come to the conclusion how exactly create/edit should work. I'm leaving this part for the future.

Because this pr completes only a part of #6, I'm not closing this issue.

### CODE REVIEW

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. "Delete provider" is not translated

image


2. Disabled buttons should has no changing color, when the user hovers on them - they should be... disabled

File: app/Http/Controllers/Admin/ProviderController.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+class ProviderController extends Controller
+{
+    public function index(): Response
+    {
+        $providers = Provider::query()
+            ->search("name")
+            ->orderByName()
+            ->orderByTimeRange()
+            ->paginate(15)
```
Comment (EwelinaSkrzypacz) [inline app/Http/Controllers/Admin/ProviderController.php:24]: Maybe we should use some const here?

File: app/Http/Controllers/Admin/ProviderController.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public function destroy(Provider $provider): void
+    {
+        $provider->delete();
+        $imagePath = storage_path("app/public/providers/" . strtolower($provider["name"]) . ".png");
+        File::delete($imagePath);
+    }
+
+    public function showLogo(string $filename): \Illuminate\Http\Response
```
Comment (EwelinaSkrzypacz) [inline app/Http/Controllers/Admin/ProviderController.php:66]: Typing

File: resources/js/Pages/Providers/Index.vue [hunk 3]
```diff
@@ -0,0 +1,216 @@
+<script setup>
+import Provider from '../../Shared/Components/Provider.vue'
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Providers/Index.vue:2]: Use `@/...` here

File: resources/js/Pages/Providers/Index.vue [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import { MagnifyingGlassIcon, XMarkIcon } from '@heroicons/vue/24/outline'
+import ErrorMessage from '@/Shared/Components/ErrorMessage.vue'
+import { onClickOutside } from '@vueuse/core'
+import { debounce } from 'lodash/function'
+import Pagination from '@/Shared/Components/Pagination.vue'
+import PaginationInfo from '@/Shared/Components/PaginationInfo.vue'
+import PrimarySaveButton from '@/Shared/Components/PrimarySaveButton.vue'
+import { useToast } from 'vue-toastification'
+import { __ } from '@/translate'
+import UploadFileButton from '../../Shared/Components/UploadFileButton.vue'
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Providers/Index.vue:15]: Use `@/...` here

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. Mouse cursor is changes, when I hover on button "Edytuj" - it' shouldn't be like that.

image

2. I don't like this hover on URL 

image

3. When I click on admin navigation "Statistics" I have 404 error.

File: resources/js/Pages/Providers/Index.vue (hunk 3 above)
Comment (AleksandraKozubal) [reply resources/js/Pages/Providers/Index.vue]: used

File: resources/js/Pages/Providers/Index.vue (hunk 4 above)
Comment (AleksandraKozubal) [reply resources/js/Pages/Providers/Index.vue]: used

File: app/Http/Controllers/Admin/ProviderController.php (hunk 2 above)
Comment (AleksandraKozubal) [reply app/Http/Controllers/Admin/ProviderController.php]: changed

File: app/Http/Controllers/Admin/ProviderController.php (hunk 1 above)
Comment (AleksandraKozubal) [reply app/Http/Controllers/Admin/ProviderController.php]: added const

**Review by EwelinaSkrzypacz — APPROVED** [review]

File: app/Models/Provider.php [hunk 5]
```diff
@@ -16,9 +18,19 @@ class Provider extends Model
...
+        "url",
+        "color",
+    ];
 
-    public function cityProvider()
+    public static function query(): Builder
     {
-        return $this->belongsTo(CityProvider::class);
+        return parent::query();
+    }
```
Comment (kamilpiech97) [inline app/Models/Provider.php:30]: Do we need this?

File: app/Http/Requests/ProviderRequest.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    public function rules(): array
+    {
+        return [
+            "name" => ["required", "string", "regex:/^[A-Z\s]/", "max:100", $this->uniqueRuleForProvider("name")],
+            "color" => ["required", "string", "size:7"],
+            "file" => [
+                "required",
+                "mimes:png",
+                File::image()
+                    ->max(10 * 1024)
```
Comment (kamilpiech97) [inline app/Http/Requests/ProviderRequest.php:23]: Lest move that to some config.

File: app/Http/Controllers/Admin/ProviderController.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        $fileContents = $request->file("file")->get();
+
+        Storage::disk("public")->put("providers/" . $fileName, $fileContents);
+    }
+
+    public function update(ProviderRequest $request, Provider $provider): void
+    {
+        $provider->update($request->validated());
+
+        $imageName = strtolower($request["name"]) . "." . $request->file("file")->getClientOriginalExtension();
```
Comment (kamilpiech97) [inline app/Http/Controllers/Admin/ProviderController.php:49]: Lest move that: `strtolower($request["name"]) . "." . $request->file("file")->getClientOriginalExtension();` to some common function in controller or helper for global use (and update other places with that).

File: app/Models/Provider.php (hunk 5 above)
Comment (AleksandraKozubal) [reply app/Models/Provider.php:30]: It's a change made in different commit, added when merging main to this branch. I do not use it, but it is used somewhere, for sure.

File: app/Http/Requests/ProviderRequest.php (hunk 6 above)
Comment (AleksandraKozubal) [reply app/Http/Requests/ProviderRequest.php]: moved

File: app/Http/Controllers/Admin/ProviderController.php (hunk 7 above)
Comment (AleksandraKozubal) [reply app/Http/Controllers/Admin/ProviderController.php]: moved to getFileName()

**Review by kamilpiech97 — APPROVED** [review]

**Review by Lee0z — APPROVED** [review]

**Pull request merged**